Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From sagar.abhishek@gmail.com Tue May 27 11:54:47 2008
Date: Sun, 25 May 2008 00:10:04 +0530
From: Abhishek Sagar <sagar.abhishek@gmail.com>
To: Ingo Molnar <mingo@elte.hu>, rostedt@goodmis.org
Cc: LKML <linux-kernel@vger.kernel.org>
Subject: [PATCH] ftrace: fix updating of ftrace_update_cnt

Hi Ingo/Steven,

Ftrace currently maintains an update count which includes false updates, i.e, updates which failed. If anything, such failures should be tracked by some separate variable, but this patch provides a minimal fix.

Signed-off-by: Abhishek Sagar <sagar.abhishek@gmail.com>
---
fix updating of ftrace_update_cnt

---
 kernel/trace/ftrace.c |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

Index: linux-2.6.25.4-rt4/kernel/trace/ftrace.c
===================================================================
--- linux-2.6.25.4-rt4.orig/kernel/trace/ftrace.c	2008-05-29 09:46:05.000000000 -0400
+++ linux-2.6.25.4-rt4/kernel/trace/ftrace.c	2008-05-29 09:46:06.000000000 -0400
@@ -452,7 +452,7 @@ static void ftrace_shutdown_replenish(vo
 	ftrace_pages->next = (void *)get_zeroed_page(GFP_KERNEL);
 }
 
-static void
+static int
 ftrace_code_disable(struct dyn_ftrace *rec)
 {
 	unsigned long ip;
@@ -468,7 +468,9 @@ ftrace_code_disable(struct dyn_ftrace *r
 	if (failed) {
 		rec->flags |= FTRACE_FL_FAILED;
 		ftrace_free_rec(rec);
+		return 0;
 	}
+	return 1;
 }
 
 static int __ftrace_update_code(void *ignore);
@@ -643,8 +645,8 @@ static int __ftrace_update_code(void *ig
 
 		/* all CPUS are stopped, we are safe to modify code */
 		hlist_for_each_entry(p, t, &head, node) {
-			ftrace_code_disable(p);
-			ftrace_update_cnt++;
+			if (ftrace_code_disable(p))
+				ftrace_update_cnt++;
 		}
 
 	}
